simple_functorch_modes.py

```diff
@@ -0,0 +1,1302 @@
+            result_names = f"{result_names[0]},"
+        else:
+            result_names = ", ".join(result_names)
+
+    print(
+        f"{result_names} = custom_vjp({fwd_fn.__name__}, {bwd_fn.__name__}, {arg_names})"
+    )
+
+class Logger(Dispatcher):
+    def __init__(self, *, name):
+        self.name = f"  {name}"
+
+    def size(self, input):
+        # don't log size calls
+        return self.inner.size(input)
+
+    def ones(self, size, **kwargs):
+        r = torch.ones(size)
+        print(f"{self.name} {r.t_name}: {self.size(r)} = ones({size})")
+        return r
+
+    def mul(self, lhs, rhs):
+        r = lhs.mul(rhs)
+        if isinstance(rhs, float):
+            print(f"{self.name} {r.t_name}: {self.size(r)} = {lhs.t_name} * {rhs}")
+        else:
+            print(
+                f"{self.name} {r.t_name}: {self.size(r)} = {lhs.t_name} * {rhs.t_name}"
+            )
+        return r
+
+    def add(self, lhs, rhs):
+        r = lhs.add(rhs)
+        print(f"{self.name} {r.t_name}: {self.size(r)} = {lhs.t_name} + {rhs.t_name}")
+        return r
+
+    def sum(self, input, dim=None):
+        if dim is None:
+          r = input.sum()
+        else:
+          r = input.sum(dim)
+        print(f"{self.name} {r.t_name}: {self.size(r)} = {input.t_name}.sum(dim={dim})")
+        return r
+
+    def unsqueeze(self, input, dim):
+        r = input.unsqueeze(dim)
+        print(
+            f"{self.name} {r.t_name}: {self.size(r)} = {input.t_name}.unsqueeze({dim})"
+        )
+        return r
+
+    def squeeze(self, input, dim):
+        r = input.squeeze(dim)
+        print(f"{self.name} {r.t_name}: {self.size(r)} = {input.t_name}.squeeze({dim})")
+        return r
+
+    def expand(self, input, sizes):
+        r = input.expand(sizes)
+        print(
+            f"{self.name} {r.t_name}: {self.size(r)} = {input.t_name}.expand({sizes})"
+        )
+        return r
+
+    def custom_vjp(self, fwd_fn, bwd_fn, *args):
+        # because custom_vjp is not an aten function, we have to explicitly send
+        # it to its inner 
+        r = super().custom_vjp(fwd_fn, bwd_fn, *args)
+        print(custom_vjp_str(r, fwd_fn, bwd_fn, args))
+        return r
+    
+    def lift(self, input, d):
+        if self == d:
+            return input
+        else:
+            return self.inner.lift(input, d)
+
+"""Here is a simple example of using Logger and Torch together.  Whenever
+we make calls to operations, we must do so via the Dispatcher object.
+We will explicitly write out all of these calls before we add wrapper
+class sugaring.
+"""
+
+with Labeler():
+    with Logger(name="Torch"):
+      z = torch.ones(2) + torch.ones(2)
+print(z)
+assert(isinstance(z, torch.Tensor))
+
+"""With the Dispatcher structure in hand, we are now in a good place to
+port the autograd implementation from Simple Autograd into our new
+framework.
+"""
+
+from torch.utils._mode_utils import no_dispatch
+
+class Autograd(Dispatcher):
```

Comment:
Note that because we're sitting at the torch dispatch level, autograd has already happened so we still have to rewrite autograd here and can't reuse PyTorch's autograd. This would be a no-go in actual functorch so we'll need to figure out a solution here